### PR TITLE
Added Native Player Gesture for Volume & Brightness 

### LINKED ISSUE #13: Add gesture controls for volume and brightness

### PR DESCRIPTION
**Pull request by yashx**
closes #13
### How it looks
gesture2

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
Can we add an option to disable this behavior?

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 1]
```diff
@@ -232,9 +247,77 @@ class PlayerActivity : AppCompatActivity() {
...
+                    overlayBrightnessVolumeProgressBar.max = maxVolume
+                    overlayBrightnessVolumeProgressBar.progress = toSet
+                }
+                else{
+                    // On left so changing brightness
+                    val windowLayoutParams = window.attributes
+                    if(counter == -1f){
+                        counter = windowLayoutParams.screenBrightness
+                        if(counter < 0f)
+                            counter = android.provider.Settings.System.getFloat(
```
Comment (yashx) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:292]: @Maxr1998 @nielsvanvelzen This line has a few problems. According to the docs, it should give value between 0 and 255 but on my phone (which has MIUI) it is goes over 255.
Worst case scenario, it jumps to full brightness first time the gesture is used.
I can disable it but imo the utility it offers is worth dealing with the bug.

**Comment by yashx** [comment]
> Can we add an option to disable this behavior?
Should I do this in this PR or send another one?

**Comment by dkanada** [comment]
Within this one should be fine.

**Comment by yashx** [comment]
@nielsvanvelzen Added option to disable it in "Client Settings"

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 2]
```diff
@@ -232,9 +253,80 @@ class PlayerActivity : AppCompatActivity() {
...
+                if (e1 == null || isBrightnessVolumeGestureAllowed.not())
+                    return false
+
+                if (abs(distanceY / distanceX) < 2)
+                    return true
+
+                val viewCenterX = playerView.measuredWidth / 2
+
+                // Distance to swipe to go from min to max
+                val distanceFull = playerView.measuredHeight
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:272]: I think 2/3 would be enough for that..

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 3]
```diff
@@ -232,9 +253,80 @@ class PlayerActivity : AppCompatActivity() {
...
+
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:273]:
```suggestion
```

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 4]
```diff
@@ -232,9 +253,80 @@ class PlayerActivity : AppCompatActivity() {
...
+                    overlayBrightnessVolumeImage.setImageResource(R.drawable.ic_volume_white_84dp)
+                    overlayBrightnessVolumeProgressBar.max = maxVolume
+                    overlayBrightnessVolumeProgressBar.progress = toSet
+                } else {
+                    // On left so changing brightness
+                    val windowLayoutParams = window.attributes
+                    if (counter == -1f) {
+                        counter = windowLayoutParams.screenBrightness
+                        if (counter < 0f)
+                            counter = android.provider.Settings.System.getFloat(
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:297]: Please add an import (alias) for either Settings or System, if possible

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 5]
```diff
@@ -34,7 +39,16 @@ class PlayerActivity : AppCompatActivity() {
     private val titleTextView: TextView by lazyView(R.id.track_title)
     private val fullscreenSwitcher: ImageButton by lazyView(R.id.fullscreen_switcher)
     private val unlockScreenButton: ImageButton by lazyView(R.id.unlock_screen_button)
+    private val overlayBrightnessVolumeLayout: LinearLayout by lazyView(R.id.overlay_b_v_layout)
+    private val overlayBrightnessVolumeImage: ImageView by lazyView(R.id.overlay_b_v_image)
+    private val overlayBrightnessVolumeProgressBar: ProgressBar by lazyView(R.id.overlay_b_v_progress_bar)
     private lateinit var playbackMenus: PlaybackMenus
+    private val audioManager: AudioManager by lazy { (getSystemService(Context.AUDIO_SERVICE) as AudioManager) }
+    private var counter = -1f
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:47]: This should be a bit more descriptive

File: app/src/main/res/drawable/ic_brightness_white_84dp.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="84dp" android:tint="#FFFFFF"
+    android:viewportHeight="24" android:viewportWidth="24"
+    android:width="84dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="@android:color/white" android:pathData="M20,8.69L20,4h-4.69L12,0.69 8.69,4L4,4v4.69L0.69,12 4,15.31L4,20h4.69L12,23.31 15.31,20L20,20v-4.69L23.31,12 20,8.69zM12,18c-3.31,0 -6,-2.69 -6,-6s2.69,-6 6,-6 6,2.69 6,6 -2.69,6 -6,6zM12,8c-2.21,0 -4,1.79 -4,4s1.79,4 4,4 4,-1.79 4,-4 -1.79,-4 -4,-4z"/>
+</vector>
```
Comment (Maxr1998) [inline app/src/main/res/drawable/ic_brightness_white_84dp.xml:5]: Please reformat the file in Android Studio

File: app/src/main/res/drawable/ic_volume_white_84dp.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="84dp" android:tint="#FFFFFF"
+    android:viewportHeight="24" android:viewportWidth="24"
+    android:width="84dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="@android:color/white" android:pathData="M3,9v6h4l5,5L12,4L7,9L3,9zM16.5,12c0,-1.77 -1.02,-3.29 -2.5,-4.03v8.05c1.48,-0.73 2.5,-2.25 2.5,-4.02zM14,3.23v2.06c2.89,0.86 5,3.54 5,6.71s-2.11,5.85 -5,6.71v2.06c4.01,-0.91 7,-4.49 7,-8.77s-2.99,-7.86 -7,-8.77z"/>
+</vector>
```
Comment (Maxr1998) [inline app/src/main/res/drawable/ic_volume_white_84dp.xml:5]: Please reformat the file in Android Studio

File: app/src/main/res/layout/activity_player.xml [hunk 8]
```diff
@@ -16,6 +16,11 @@
         android:layout_width="match_parent"
         android:layout_height="match_parent">
 
+        <include
+            layout="@layout/overlay_brightness_volume"
+            android:visibility="gone"
+            tools:visibility="visible"/>
```
Comment (Maxr1998) [inline app/src/main/res/layout/activity_player.xml:22]:
```suggestion
            tools:visibility="visible" />
```

**Review by Maxr1998 — CHANGES_REQUESTED** [review]
Code looks mostly good, I have a few comments about naming and formatting stuff however. I'd generally use a shorter name for the gesture feature.
Also, I'm not a fan of that `counter` attribute.. I'm not sure how you could do it differently though, I think I'll have to play with the code myself once the other mentioned issues are fixed.

**Comment by yashx** [comment]
@Maxr1998 While brightness gesture can be implemented without the counter. The problem is with the volume gesture. I can also set volume by using the distance between  `e1` and `e2` in `onScroll` but I would need to store the initial volume level when the gesture was started.
So I decided to go with the `counter` approach so that all gestures  (brightness, volume and horizontal seek(future PR)) can have a similar implementation.

**Comment by Maxr1998** [comment]
I see, then there's probably no way around that. Could you still make the counter variable a bit more descriptive then, and add some documentation/comments on how you're using it?

**Comment by yashx** [comment]
Yes I will do that and all the requested changes in a few hours.

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 9]
```diff
@@ -232,9 +261,72 @@ class PlayerActivity : AppCompatActivity() {
                 }
                 return true
             }
+
+            override fun onScroll(e1: MotionEvent?, e2: MotionEvent?, distanceX: Float, distanceY: Float): Boolean {
+                if (e1 == null || !isBrightnessVolumeGestureAllowed)
+                    return false
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:267]:
```suggestion
                e1 as? isBrightnessVolumeGestureAllowed ?: return false
```

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 10]
```diff
@@ -232,9 +261,72 @@ class PlayerActivity : AppCompatActivity() {
                 }
                 return true
             }
+
+            override fun onScroll(e1: MotionEvent?, e2: MotionEvent?, distanceX: Float, distanceY: Float): Boolean {
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:265]: Please change `e1` and `e2` to more descriptive names.

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt (hunk 9 above)
Comment (yashx) [reply app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:267]: Uhm isn't this code incorrect? I am new to kotlin so correct me if I am wrong but why should I cast a MotionEvent object to a Boolean object?

image

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt (hunk 10 above)
Comment (yashx) [reply app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:265]: `e1` and `e2` are the names used in the documentation of `onScroll` method. Are you sure I should change them?

image

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt (hunk 9 above)
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:267]: Sorry I read the code wrong due too GitHub's syntax highlighting. Ignore this one!

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt (hunk 10 above)
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:265]: I think something like `firstMotionEvent` and `secondMotionEvent` is more clear. @Maxr1998 what do you think?

Comment (Maxr1998) [reply app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:265]: Maybe `first` and `current`? Imo no need to state the type in the variable. And even if, then only append `Event`, not the full name.

**Review by nielsvanvelzen — APPROVED** [review]

**Comment by Maxr1998** [comment]
@yashx did some more tweaks and code style improvements, please let me know if it's ok for you, and if you @nielsvanvelzen also don't have any comments anymore, I'll squash and merge it.

**Review by nielsvanvelzen — APPROVED** [review]

**Review by Maxr1998 — APPROVED** [review]

**Pull request merged**